### PR TITLE
Multi-level RNN example

### PR DESCRIPTION
**Pull request by pasky**
Multi-level RNNs are getting more popular, important especially (but not only) for character level models. Moreover, they are a bit unintuitive in Keras when they are bidirectional. This example demonstrates how to construct them, a few common tricks, and a few exercises that might prove educational for newcomers to RNNs for NLP.

I had to answer a question about how to build these in Keras maybe four times in the last month alone, I really see a lot of interest and the github issues of Keras contain scattered, obsolete and misleading information.  I also tried to point out some common pitfalls, e.g. not taking result variance into account.

### CODE REVIEW

File: examples/imdb_multilevel_gru.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+max_features = 20000
+maxlen = 100  # cut texts after this number of words (among top max_features most common words)
+batch_size = 32
+N = 128  # base dimensionality of layers
+rnn_levels = 2
+
+
+def reverse_time(x):
+    import keras.backend as K
+    assert K.ndim(x) == 3, "Should be a 3D tensor."
```
Comment (fchollet) [inline examples/imdb_multilevel_gru.py]: Nit: mixing `'` and `"` in this file; normalize all to `'`.

File: examples/imdb_multilevel_gru.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+max_features = 20000
+maxlen = 100  # cut texts after this number of words (among top max_features most common words)
+batch_size = 32
+N = 128  # base dimensionality of layers
+rnn_levels = 2
+
+
+def reverse_time(x):
+    import keras.backend as K
```
Comment (fchollet) [inline examples/imdb_multilevel_gru.py]: Imports should be at the top.

File: examples/imdb_multilevel_gru.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+print("Pad sequences (samples x time)")
+X_train = sequence.pad_sequences(X_train, maxlen=maxlen)
+X_test = sequence.pad_sequences(X_test, maxlen=maxlen)
+print('X_train shape:', X_train.shape)
+print('X_test shape:', X_test.shape)
+y_train = np.array(y_train)
+y_test = np.array(y_test)
+
+
```
Comment (fchollet) [inline examples/imdb_multilevel_gru.py]: Unnecessary whitespace.

File: examples/imdb_multilevel_gru.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+# this is the placeholder tensor for the input sequences
+sequence = Input(shape=(maxlen,), dtype='int32')
+# this embedding layer will transform the sequences of integers
+# into vectors of size 128
+embedded_seq = Embedding(max_features, N, input_length=maxlen)(sequence)
+
+
```
Comment (fchollet) [inline examples/imdb_multilevel_gru.py]: Unnecessary whitespace.

File: examples/imdb_multilevel_gru.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    # apply forwards GRU
+    forwards = GRU(N, return_sequences=True)(rnn_input)
+    # apply backwards GRU
+    backwards = GRU(N, return_sequences=True, go_backwards=True)(rnn_input)
+
+    # the backwards sequence is in reverse order, so we need to flip it
+    # if we are to element-wise merge it with forwards!
+    backwards = Lambda(reverse_time)(backwards)
+
+
```
Comment (fchollet) [inline examples/imdb_multilevel_gru.py]: Unnecessary whitespace.

File: examples/imdb_multilevel_gru.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+#   forwards, first token for backwards).  This is shown here and
+#   the same as if we set return_sequences=False for the last GRU level.
+# * MaxPooling1D with pool_length=maxlen (i.e. element-wise max over
+#   the sequence)
+# * AveragePooling1D or other variations.
+forwards1 = Lambda(time_last, output_shape=time_one_shape)(forwards)
+backwards1 = Lambda(time_first, output_shape=time_one_shape)(backwards)
+aggregate = merge([forwards1, backwards1], mode='sum')
+
+
```
Comment (fchollet) [inline examples/imdb_multilevel_gru.py]: Unnecessary whitespace.

**Comment by fchollet** [comment]
The docstring would need more detailed explanation of how the model works (this will be read by people with no previous background in bi-directional RNNs).

How about instead creating a PR for a `Bidirectional` wrapper for RNNs, in `layers/wrappers.py`? This would be more reusable and would be readily applicable to any existing RNN. You already have all the code to make it happen ;-)

**Comment by kemaswill** [comment]
It seems that the simple "shallow" model such as [CNN](http://www.aclweb.org/anthology/D14-1181), [FastText](https://arxiv.org/abs/1607.01759) can achieve quite good performance on text classification w.r.t. both accuracy and training time. Could you explain why should we apply the multi-level RNN on text classification? Any paper reference will be appreciated.

**Comment by fchollet** [comment]
@pasky : we have merged the `Bidirectional` wrapper. What do you want to do with your example -- do you want to adapt it to use the `Bidirectional` wrapper?

**Comment by farizrahman4u** [comment]
@pasky You can make your code much prettier with the `Bidirectional` wrapper : 

```python
input = Input((maxlen,))
embeddings = Embedding(max_features, 128)(input)
rnn_output = embeddings
for i in range(nb_levels):
	rnn_output = Dropout(0.3)(rnn_output)
	rnn_output = Bidirectional((GRU(128, return_sequences=True)), merge_mode='sum')(rnn_output)
	rnn_output = merge([rnn_output, embeddings], mode='sum')  # Feeding the input sequence at every level imporves accuracy
output = Lambda(lambda x:x[:, -1, :], output_shape=lambda s: (s[0], s[2]))(rnn_output)  # We require the final timestep only
output = Dropout(0.25)(output)
output = Dense(1, activation='sigmoid')(output)
model = Model(input=input, output=output)
```

**Comment by pasky** [comment]
Oh, I'm glad that Bidirectional() layer indeed did land meanwhile! I think that makes the example less educational, especially since the accuracy improvement is very minuscule.

One general class of problems not included in Keras examples is sequence tagging, and multi-level RNNs shine in that area a lot. The problem is that e.g. the CoNLL2003 NER dataset (which is the most standard sequence tagging dataset I'd say) is not freely available. My past models were run on non-public data too.

Any ideas re what data we could use to demo this? I'll think about it too, and be happy to convert the example to a sequence prediction task, which I think would make it a lot more useful.

**Comment by farizrahman4u** [comment]
There should be some open sequence tagging dataset.. it doesn't seem sensible to use this for imdb since fasttext example can reach 90% accuracy in 3 less than seconds.

**Comment by pasky** [comment]
So my restrictions are: (A) freely available, ideally easily automatically downloadable from 3rd-party location; (B) English if possible (I could use the opportunity to also show how to preinitialize word embeddings from GloVe or such.); (C) ideally referenced in some recent academic papers.

I looked at part-of-speech tagging and treebanks, but the free ones all seem kind of obscure.

A NER option may be http://schwa.org/projects/resources/wiki/Wikiner though it's a bit obscure (didn't see a paper from other people than its authors that would benchmark on it).

Some bleeding edge datasets for NLP sequence tagging that come to mind are the WikiReading dataset and https://stanford-qa.com/ - but the task requires some explaining, is pretty hard and the datasets aren't that small. But we could just go with that anyway...

**Comment by fchollet** [comment]
I would consider using Stanford QA since it will only rise in popularity over the next couple years. It will be good to have a `stanford_qa` module in `keras.datasets` to automatically download and process it.

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.
